cdap-watchdog/src/main/java/io/cdap/cdap/logging/gateway/handlers/RemoteProgramLogsFetcher.java

```diff
@@ -0,0 +1,127 @@
+ * the License.
+ */
+
+package io.cdap.cdap.logging.gateway.handlers;
+
+import com.google.common.base.Charsets;
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import io.cdap.cdap.common.NotFoundException;
+import io.cdap.cdap.common.ProgramNotFoundException;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.conf.Constants.Gateway;
+import io.cdap.cdap.common.http.DefaultHttpRequestConfig;
+import io.cdap.cdap.common.internal.remote.RemoteClient;
+import io.cdap.cdap.common.internal.remote.RemoteClientFactory;
+import io.cdap.cdap.internal.app.store.RunRecordDetail;
+import io.cdap.cdap.proto.id.ProgramId;
+import io.cdap.cdap.security.spi.authentication.UnauthenticatedException;
+import io.cdap.cdap.security.spi.authorization.UnauthorizedException;
+import io.cdap.common.http.HttpMethod;
+import io.cdap.common.http.HttpRequest;
+import io.cdap.common.http.HttpResponse;
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+
+/**
+ * Fetch Program logs via internal REST API calls
+ */
+public class RemoteProgramLogsFetcher implements ProgramLogsFetcher {
+
+  private final RemoteClient remoteClient;
+
+  @Inject
+  public RemoteProgramLogsFetcher(RemoteClientFactory remoteClientFactory) {
+    this.remoteClient = remoteClientFactory.createRemoteClient(
+        Constants.Service.LOG_QUERY,
+        new DefaultHttpRequestConfig(false),
+        Gateway.API_VERSION_3);
+  }
+
+  /**
+   * Gets the run logs of a program.
+   *
+   * @param program the program
+   * @param runId   pipeline run id
+   * @param start   start time of the time range of desired logs
+   * @param stop    end time of the time range of desired logs
+   * @return the logs of the program
+   * @throws IOException              if a network error occurred
+   * @throws NotFoundException        if the application or program could not be found
+   * @throws UnauthenticatedException if the request is not authorized successfully in the gateway
+   *                                  server
+   */
+  @Override
+  public String getProgramRunLogs(ProgramId program, String runId, long start, long stop)
+      throws IOException, NotFoundException, UnauthenticatedException, UnauthorizedException {
+
+    String url = String.format("namespaces/%s/apps/%s/%s/%s/runs/%s/logs?start=%d&stop=%d",
+                               program.getNamespaceId().getNamespace(),
+                               program.getApplication(), program.getType().getCategoryName(),
+                               program.getProgram(), runId, start, stop);
+    HttpRequest.Builder requestBuilder = remoteClient.requestBuilder(HttpMethod.GET, url);
+    HttpResponse response;
+    response = execute(requestBuilder.build());
+    if (response.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+      throw new ProgramNotFoundException(program);
+    }
+
+    return new String(response.getResponseBody(), Charsets.UTF_8);
```

Comment:
Are we keeping the whole logs in memory? This can be huge and goes OOM. Any reading and writing of logs should be streamed.